execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -925,52 +890,67 @@ where
     S::Error: From<T::Error>,
     E: From<S::Error> + From<bytesrepr::Error>,
 {
-    match store.get(txn, root)? {
+    match store.get_raw(txn, root)? {
         None => Ok(WriteResult::RootNotFound),
-        Some(current_root) => {
+        Some(current_root_bytes) => {
             let new_leaf = Trie::Leaf {
                 key: key.to_owned(),
                 value: value.to_owned(),
             };
             let path: Vec<u8> = key.to_bytes()?;
-            let TrieScan { tip, parents } =
-                scan::<K, V, T, S, E>(txn, store, &path, &current_root)?;
+            let TrieScanRaw { tip, parents } =
+                scan_raw::<K, V, T, S, E>(txn, store, &path, current_root_bytes)?;
             let new_elements: Vec<(Digest, Trie<K, V>)> = match tip {
-                // If the "tip" is the same as the new leaf, then the leaf
-                // is already in the Trie.
-                Trie::Leaf { .. } if new_leaf == tip => Vec::new(),
-                // If the "tip" is an existing leaf with the same key as the
-                // new leaf, but the existing leaf and new leaf have different
-                // values, then we are in the situation where we are "updating"
-                // an existing leaf.
-                Trie::Leaf {
-                    key: ref leaf_key,
-                    value: ref leaf_value,
-                } if key == leaf_key && value != leaf_value => rehash(new_leaf, parents)?,
-                // If the "tip" is an existing leaf with a different key than
-                // the new leaf, then we are in a situation where the new leaf
-                // shares some common prefix with the existing leaf.
-                Trie::Leaf {
-                    key: ref existing_leaf_key,
-                    ..
-                } if key != existing_leaf_key => {
-                    let existing_leaf_path = existing_leaf_key.to_bytes()?;
-                    let (new_node, parents) = reparent_leaf(&path, &existing_leaf_path, parents)?;
-                    let parents = add_node_to_parents(&path, new_node, parents);
-                    rehash(new_leaf, parents)?
+                LazyTrieLeaf::Left(leaf_bytes) => {
+                    let trie_tag = trie::lazy_trie_tag(leaf_bytes.as_slice());
+                    assert_eq!(
+                        trie_tag,
+                        Some(TrieTag::Leaf),
+                        "Unexpected trie variant found instead of a `TrieTag::Leaf`"
+                    );
```

Comment:
Are we sure we want to panic inside `write()` in case of mismatch?

Reply:
Currently, it's guaranteed by lazy trie deserializer that the panic wouldn't happen - we have ticket #3994 to address this by using proper enum rather than Either